File: frontend/src/api/accounts.ts
import apiClient from './client';
import type { Account, AccountCreate, AccountUpdate } from '../types';

export async function getAccounts(scenarioId?: number): Promise<Account[]> {
  const params = scenarioId !== undefined ? { scenario_id: scenarioId } : {};
  const res = await apiClient.get<Account[]>('/accounts/', { params });
  return res.data;
}

export async function getAccount(id: number): Promise<Account> {
  const res = await apiClient.get<Account>(`/accounts/${id}`);
  return res.data;
}

export async function createAccount(data: AccountCreate): Promise<Account> {
  const res = await apiClient.post<Account>('/accounts/', data);
  return res.data;
}

export async function updateAccount(id: number, data: AccountUpdate): Promise<Account> {
  const res = await apiClient.put<Account>(`/accounts/${id}`, data);
  return res.data;
}

export async function deleteAccount(id: number): Promise<void> {
  await apiClient.delete(`/accounts/${id}`);
}

File: backend/app/routers/accounts.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from typing import List

from app.database import get_db
from app.models.models import Account
from app.schemas.schemas import AccountCreate, AccountRead, AccountUpdate

router = APIRouter(prefix="/accounts", tags=["accounts"])


@router.get("/", response_model=List[AccountRead])
def list_accounts(scenario_id: int = None, db: Session = Depends(get_db)):
    q = db.query(Account)
    if scenario_id is not None:
        q = q.filter(Account.scenario_id == scenario_id)
    return q.order_by(Account.withdrawal_priority).all()


@router.get("/{account_id}", response_model=AccountRead)
def get_account(account_id: int, db: Session = Depends(get_db)):
    account = db.query(Account).filter(Account.id == account_id).first()
    if not account:
        raise HTTPException(status_code=404, detail="Account not found")
    return account


@router.post("/", response_model=AccountRead, status_code=201)
def create_account(data: AccountCreate, db: Session = Depends(get_db)):
    account = Account(**data.model_dump())
    db.add(account)
    db.commit()
    db.refresh(account)
    return account


@router.put("/{account_id}", response_model=AccountRead)
def update_account(account_id: int, data: AccountUpdate, db: Session = Depends(get_db)):
    account = db.query(Account).filter(Account.id == account_id).first()
    if not account:
        raise HTTPException(status_code=404, detail="Account not found")
    for field, value in data.model_dump(exclude_unset=True).items():
        setattr(account, field, value)
    db.commit()
    db.refresh(account)
    return account


@router.delete("/{account_id}", status_code=204)
def delete_account(account_id: int, db: Session = Depends(get_db)):
    account = db.query(Account).filter(Account.id == account_id).first()
    if not account:
        raise HTTPException(status_code=404, detail="Account not found")
    db.delete(account)
    db.commit()
